components/datetime/src/skeleton.rs

```diff
@@ -195,6 +200,67 @@ impl TryFrom<&str> for Skeleton {
+///
+/// At the time of this writing, it's being used for applying hour cycle preferences and should not
+/// be exposed as a public API for end users.
+#[doc(hidden)]
+#[cfg(feature = "provider_transform_utils")]
+impl From<&Pattern> for Skeleton {
+    fn from(pattern: &Pattern) -> Self {
+        let mut fields: SmallVec<[fields::Field; 5]> = SmallVec::new();
+        for item in pattern.items() {
+            if let crate::pattern::PatternItem::Field(field) = item {
+                let mut field = *field;
+
+                // Skeletons only have a subset of available fields, these are then mapped to more
+                // specific fields for the patterns they expand to.
+                field.symbol = match field.symbol {
+                    FieldSymbol::Year(year) => FieldSymbol::Year(year),
+                    // Remove the stand-alone designation. This will be provided by the pattern.
+                    FieldSymbol::Month(fields::Month::Format)
+                    | FieldSymbol::Month(fields::Month::StandAlone) => {
+                        FieldSymbol::Month(fields::Month::Format)
+                    }
+                    FieldSymbol::Day(day) => FieldSymbol::Day(day),
+                    FieldSymbol::Weekday(_) => FieldSymbol::Weekday(fields::Weekday::Format),
+
+                    // Only flexible day periods are used in skeletons, ignore all others.
+                    FieldSymbol::DayPeriod(fields::DayPeriod::AmPm)
+                    | FieldSymbol::DayPeriod(fields::DayPeriod::NoonMidnight) => continue,
+                    // TODO(#487) - Flexible day periods should be included here.
+                    // FieldSymbol::DayPeriod(fields::DayPeriod::Flexible) => {
+                    //     FieldSymbol::DayPeriod(fields::DayPeriod::Flexible)
+                    // }
+
+                    // Only the H12 and H23 symbols are used in skeletons, while the patterns may
+                    // contain H11 or H23 depending on the localization.
+                    FieldSymbol::Hour(fields::Hour::H11) | FieldSymbol::Hour(fields::Hour::H12) => {
+                        FieldSymbol::Hour(fields::Hour::H12)
+                    }
+                    FieldSymbol::Hour(fields::Hour::H23) | FieldSymbol::Hour(fields::Hour::H24) => {
+                        FieldSymbol::Hour(fields::Hour::H23)
+                    }
+
+                    // Pass through all of the following preferences unchanged.
+                    FieldSymbol::Minute => FieldSymbol::Minute,
+                    FieldSymbol::Second(second) => FieldSymbol::Second(second),
+                    FieldSymbol::TimeZone(time_zone) => FieldSymbol::TimeZone(time_zone),
+                };
+
+                // Only insert if it's a unique field.
+                if let Err(pos) = fields.binary_search(&field) {
+                    fields.insert(pos, field)
+                }
+            }
+        }
+        Self(fields)
+    }
+}
```

Comment:
`suggestion (n-b)`: You can make maintenance of that logic a bit easier if you make more sweeping assumptions. Something like:

```rust
for let item in pattern.items() {
    if let PatternItem::Field(field) = item {
        let new_field = match field.symbol {
            FieldSymbol::Month(_) => FieldSymbol::Month(fields::Month::Format),
            FieldSymbol::Weekday(_) => FieldSymbol::Weekday(fields::Weekday::Format),
            FieldSymbol::Hour(hour) => match hour {
                Hour::H11 | Hour::H12 => FieldSymbol::Hour(Hour::H12),
                Hour::H23 | Hour::H24 => FieldSymbol::Hour(Hour::H23),
            }

            FieldSymbol::Year(_) | FieldSymbol::Day(_) | FieldSymbol::Weekday(_) |
                FieldSymbol::Minute | FieldSymbol::Second(_) | FieldSymbol::TimeZone(_) => *field.symbol,
        };
        ...
    }
}
```

Reply:
`question`: where do you sort?

Reply:
> is this intended for runtime

No, it's only used as a provider transform, and it is explicitly marked as so with the feature `provider_transform_internals`, so it's not perf-sensitive. I'm going to pass on the perf suggestions, as I believe the current implementation is good enough, and I would posit simpler for having one step rather than 3.

> You can make maintenance of that logic a bit easier if you make more sweeping assumptions.

I tweaked things a bit to follow your suggestion here. My process for writing things was pretty organic, but this should make the matching a bit cleaner.

> where do you sort?